File: 0141-linked-list-cycle/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* here=head;
        ListNode* tortoise=head;
        bool ans=false;
        while(1){
            if(here==nullptr||here->next==nullptr||here->next->next==nullptr)break;
            here=here->next->next;
            tortoise=tortoise->next;
            if(here==tortoise){
                ans=true;
                break;
            }
        }
        return ans;
    }
};

File: 0141-linked-list-cycle/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        if(head == null) return false;
        ListNode slow = head;
        ListNode fast = head.next;
        while(fast != null && slow != null){
            if(fast == slow) return true;
            if(fast.next != null) fast = fast.next.next;
            else fast = null;
            slow = slow.next;
        }
        return false;
    }
}
